### DISCUSSION TITLE
The Future of Vyxal

### DISCUSSION THREAD

**Discussion by hyperneutrino** [Vyxal/Vyxal#191] [Q&A] [answered]
Since lyxal has announced a full reboot of Vyxal, it's time we take this in the right direction and approach this properly last time. Vyxal lacked direction up until this point - not as anyone's fault, but this started as lyxal's project and gained contributors and users rapidly and before we knew it, too many cooks spoiled the broth. This was especially evident in the clean-up, which ended in a massive fire, essentially, because there was no clear direction and the different directions pulled the project apart.

Here are two questions that I believe can help us discuss a better organization and structure for the future of Vyxal. This was initially inspired by me trying to get a feel for whether or not using Jira would be a good idea for a project of this scale, since it is a bit overpowered for what we need, but please answer these questions regardless of your opinion on that. The first question is for everyone, and the second is primarily for lyxal, but I would greatly appreciate your feedback to it as well.

1. What are your goals from Vyxal, or what do you want to contribute / want out of working on Vyxal?
  - Is it just to have some fun with a casual project?
  - Is it to help lyxal out with making their language?
  - Is it to get some experience (whether or not it's necessarily helpful for either future projects or future professional things)?
  - etc.

2. What do you want the Vyxal project and development team to be like?
  - more of a casual gathering, like a group of friends, working on a fun idea together?
  - an organized team working to create a project (of course that doesn't mean there's no fun or that it's anything close to a company structure - we don't need to have any hierarchy and can just be a couple of fellow CGCCers working on a thing but still be organized)? etc.

Please answer these questions so we can get a feel for the direction we want, since lyxal has expressed that they want the future of Vyxal to be a collaborative effort and for all of us to get a voice too, and that while they get the final say, their suggestions are not always the final say.

To be clear, Vyxal is _not_ becoming _anything_ like a company structure. There is no hierarchy (other than that lyxal owns the project and has full control, obviously), we don't have people reporting to other people, there isn't time logging, etc, all of that corporate B.S. That doesn't mean we can't be organized though, but these questions help us explore what the future of Vyxal development may look like.

**Comment by lyxal** [comment]
_Disclaimer: this answer is not to be taken as anything but opinion._

Answer 1:

To quote `Why.md`:

> Vyxal's raison d'être is to provide structures of practical languages - such as functions, variables and comments - that are oftentimes lost within the modern golfing language market.

Consequently, what I want with Vyxal is for it to be the most practical golfing language available. But I also want the experience of contributing to/leading an open source project bigger than just a few people.

Answer 2:

Currently, I would say that the vyxal community is laid back and casual. That's the way I like it. Having said that, a bit of structure and organisation would be welcome, as the current atmosphere has higher chances of things crashing and burning due to communication/procedures being _too_ lax, or even non-existent.

Any protocols we establish aren't to be hard and fast rules, but something we casually comply with and that we just do as a habit.

The current "hierarchy" of the community is fine, but I've been considering whether or not there should be a few more people with veto powers. That way, power isn't centralised within my decisions.

Finally, vyxal development/community interactions should be fun. The moment it stops being a pleasent experience is when things have gone too far. So long as protocols don't make things miserable for the majority, things should be poggers.

**Reply by hyperneutrino** [reply]
Thanks for the answer! I totally agree with the laid back and casual style - it's something I've liked about working with the Vyxal group too; I was able to join right in and just have a good time without pressure to get things done or figure stuff out right away. I guess the best way to explain the philosophy I have in mind is that we are organizing the code and development, not the people and environment. The protocols we make should restrict how people can contribute changes to the project and develop Vyxal, not the experience that people get out of this.

Finding the balance between organized and casual structure is definitely a challenge and hopefully everyone will be patient as that's figured out, but I feel that the most important part of it is continual feedback - no system is necessarily perfect and the system that sounds like it makes sense might be totally wrong - and as long as we are willing to change around our needs and desires, we'll have a positive environment, and I am glad that you've made the community a focus and fully support that.

Finally, I 100% second your last statement. Nothing we do should ever detract from reasonable enjoyable experiences. That is, having fun should not be infringed on by any of our policies (unless your idea of fun is "rewrite the add function to always return `69`" of course) and the experience needs to be enjoyable for everyone involved - only then can we think about how to go about the development itself.

**Reply by lyxal** [reply]
> organizing the code and development, not the people and environment

Someone make this into an inspirational quote or smth because it's good

**Reply by Underslash12** [reply]
_Short answer but . . ._

For part 1, personally, all of those things come to mind as reasons why I would want to contribute to Vyxal in the future (as I haven't so far).

For part 2, I second (third?) Lyxal and Hyper's takes.

**Reply by zoomlogo** [reply]
I agree with @Lyxal

**Comment by AviFS** [comment]
I have lots on my mind and that I'm working on these days. So won't be as active a part as I'd like to be-- don't count on me!  But here are my two cents:

1. I'm in because it's a project & community I want to be an active part of, and one I want to contribute to & help out. Honestly more so for the community than the language itself.

2. I definitely want it casual, fun & friendly but I'm confident that anything we do will be all of those things. That is, given all of you! I think the question is whether we'd benefit from some more organization, and at the point that we do, we should add it. <br><br> I can't emphasize enough how confident I am that it'll still be as fun as ever. And in some ways it'll be even more fun if we feel it's more extensible, stable, better documented, etc. So my vote is to add more organization as soon as we stand to benefit from it! I don't know enough the latest disaster and whether we're at that stage yet, but I imagine we are.

**Comment by pxeger** [comment]
> What are your goals from Vyxal, or what do you want to contribute / want out of working on Vyxal?

I started working on Vyxal because I saw its code was... suboptimal ...to maintain, and this was resulting in a lot of bugs. The cleanup was part of what I hoped would be the solution to this, except people started acting on it before any proper agreement was reached, and/or contrary to the agreement. This seems to be a running issue: changes are made before they were ready because they weren't discussed properly.

The rewrite seems to have started to address some of these things in the contributing guidelines e.g. by saying no pushing to master/whatever, but I would like to help by saying:
decisions should not be made, PRs should not be merged, issues should not be closed or perhaps even worked on, until several people are agreed and all discussion is at least somewhat finalised. Take things a bit slower!

> What do you want the Vyxal project and development team to be like?

<s>In an ideal world, I would be the malevolent dictator for life and be the only one allowed to work on it. *evil laugh*</s>

It seems mostly fine as it is; I don't think we don't need a huge amount of organisation I think as long as we have some self-restraint in waiting for feedback from other people. When it comes to larger sub-projects, ad-hoc structures can be created, but I don't think anything concrete is often necessary.

**Reply by ysthakur** [reply]
I'd like to add that every PR should request reviews from at least two people who are relevant to the PR, preferably a different person for each aspect of it if it's a big PR (which it shouldn't be, really, unless we're talking about merging Fresh-Beginnings with master).

**Comment by lyxal** [comment]
This is User's (@ysthakur's) answers ([sauce](https://github.com/Vyxal/Sandbox/blob/master/questions%20and%20answers.txt))

1. All of these, honestly. Vyxal is fun despite the frequent fires (often caused by me :P), y'all are great, and I get a little practice with Git, GitHub, testing, stuff like that
2. More of a casual gathering - we need a bit of organization, but no real hierarchy (except for lyxal basically having a veto on major decisions, since it's their language)

**Reply by ysthakur** [reply]
Thanks! I'll probably expand on that later here. Edit: I have.

I'm probably the only contributor here who's absolutely terrible at Vyxal. I also hate Python. Why am I contributing to  Vyxal, then? Because I like the idea of a readable golfing language, even if I myself haven't mastered it yet. It's also nice to get a bit of experience with using git, GitHub, PRs, etc. Most of all, though, this community is nice, and it's a nice project to work on in my spare time, with very little commitment required. I'd like to see it stay that way, and for Vyxal's group of contributors to grow bigger.

2. While I'd love for us to experiment with Jira and other stuff to make us better coordinated, if at any point it seems too "professional" or it starts sucking the joy out of working on Vyxal, I say we drop it immediately (I believe this has been said by others in this thread already). One good thing about Vyxal is that there's no pressure and no real responsibilities. If the tickets make it seem like we *have* to work on something by a deadline, they're not worth the organization.

**Reply by lyxal** [reply]
You aren't the only one terrible with Vyxal...hyper has previously confessed to stack based languages not being a strength of theirs, and people such as pxeger and avi also haven't apparently dived deep into the language either.

Being unfamiliar with the language is perfectly fine. What matters is that you're willing to contribute and that you're willing to learn where needed.

**Reply by zoomlogo** [reply]
i dont even know the language

**Comment by AMiller42** [comment]
Before I answer the questions, a little background on how I joined Vyxal. (**If you don't feel like reading it, there's a TL;DR at the bottom**) I was still really new to CGCC, having discovered it around February/March of 2021 when I was reading about polyglots. I checked it out, discovered esolangs, answered a couple questions in Befunge, learned golfscript, answered a couple questions in it, forgot about it, and decided to make my own language.

About this time, I noticed a post about something called Language of the Month. At the time, the LotM was Vim, which I had never really used all that much. I tried it out, and really enjoyed it. However, I wanted more rep on the site. I was new and felt like I wasn't getting any rep at all. So, I found the Bounties with No Deadline. There was one that seemed pretty easy. All I had to do was answer 5 questions in this poggers-looking language called Vyxal. I thought that would be a good one to do, so I checked out the documentation for the language, and that same evening I posted my 5th Vyxal answer.

While I had originally figured my time with Vyxal would be a one-off thing, I quickly changed my mind. I had always felt intimidated by these monstrous-looking golfing languages like Jelly and 05AB1E, but Vyxal was really easy to learn, as well as being very readable, and included a lot of features that I understood from functional programming languages. So, I decided to keep using it. I upvoted the LotM nomination, and before I knew it, I was golfing Vyxal answers left and right and could read Vyxal better than I could read Spanish. (Disclaimer: I don't know Spanish. At all. Not in the slightest.)

**TL;DR: I gave Vyxal a chance because of the bounty, and I never looked back.**

That brings me to the questions presented in the main post.

1. I started contributing to Vyxal because it seemed like a good way to practice/learn Python (the only other experience I had with Python was modifying an existing interpreter to make the interpreter for my language). I also contributed because I liked the language, and the small community around it, and I wanted to be a part of that community. I *kept* contributing to Vyxal because I've never had a chance to work on something like this with a group like this, and have loved having that small part in this amazing project. As far as what I want out of working on Vyxal, I want it to become the best thing it can be, and to know that I helped it become that. Other than that, I'm just having fun being part of the community. It's the journey, not the destination.

2. I agree that the project could use more direction and organization, though it's mostly good the way it is. As far as the people, other than more communication and collaboration to make sure that the grease fire that is the clean up doesn't happen again, I don't want anything to change. The community is one of the main reasons that I've stuck around with Vyxal, and I've fallen in love with it. For the most part, I think we've been doing okay with most stuff, but I agree with @pxeger. We need to slow down and make sure to do things right, especially when it comes to the big stuff, but also with the small stuff. This community is one that is built on collaboration, so let's collaborate.

**Comment by pxeger** [comment, marked as answer]
~~Do you think I'm safe to unpin and close this now?~~

Closing.
